modules/apps/segments/segments-web/src/main/java/com/liferay/segments/web/internal/display/context/EditSegmentsEntryDisplayContext.java

```diff
@@ -374,25 +376,40 @@ private JSONObject _getInitialQueryJSONObject(
 		FilterParser filterParser = _filterParserProvider.provide(
 			segmentsCriteriaContributor.getEntityModel());
 
-		Expression expression = filterParser.parse(criterionFilterString);
+		try {
+			Expression expression = filterParser.parse(criterionFilterString);
+
+			JSONObject jsonObject = (JSONObject)expression.accept(
+				new ExpressionVisitorImpl(
+					1, segmentsCriteriaContributor.getEntityModel()));
+
+			if (Validator.isNull(jsonObject.getString("groupId"))) {
+				jsonObject = JSONUtil.put(
+					"conjunctionName",
+					StringUtil.toLowerCase(
+						String.valueOf(BinaryExpression.Operation.AND))
+				).put(
+					"groupId", "group_0"
+				).put(
+					"items", JSONUtil.putAll(jsonObject)
+				);
+			}
 
-		JSONObject jsonObject = (JSONObject)expression.accept(
-			new ExpressionVisitorImpl(
-				1, segmentsCriteriaContributor.getEntityModel()));
+			return jsonObject;
+		}
+		catch (Exception exception) {
+			if (_log.isDebugEnabled()) {
+				_log.error(exception);
+			}
 
-		if (Validator.isNull(jsonObject.getString("groupId"))) {
-			jsonObject = JSONUtil.put(
-				"conjunctionName",
-				StringUtil.toLowerCase(
-					String.valueOf(BinaryExpression.Operation.AND))
-			).put(
-				"groupId", "group_0"
-			).put(
-				"items", JSONUtil.putAll(jsonObject)
-			);
+			if (!GetterUtil.getBoolean(
+					PropsUtil.get("feature.flag.LPS-166954"))) {
+
+				throw new Exception(exception);
+			}
 		}
 
-		return jsonObject;
+		return JSONUtil.put("error", StringPool.TRUE);
```

Comment:
You should send the type of error you are found, error = true, is not very descriptive for the frontend.

Reply:
Done, thank you